### PR TITLE
MQE: Add support for count aggregation

### PR DESCRIPTION
**Pull request by jhesketh**
#### Which issue(s) this PR fixes or relates to

Fixes #<issue number>

#### Checklist

- [x] Tests updated.
- [ ] Documentation added.
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [ ] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

File: pkg/streamingpromql/aggregations/count.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+		g.floatPresent, err = types.BoolSlicePool.Get(timeRange.StepCount, memoryConsumptionTracker)
+		if err != nil {
+			return err
+		}
+		g.floatValues = g.floatValues[:timeRange.StepCount]
+		g.floatPresent = g.floatPresent[:timeRange.StepCount]
+	}
+
+	for _, p := range data.Floats {
+		idx := (p.T - timeRange.StartT) / timeRange.IntervalMs
```
[the remark this answers is not in the document]
Comment (jhesketh) [reply pkg/streamingpromql/aggregations/count.go:39]: Good idea, will add it later

File: pkg/streamingpromql/aggregations/count.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+	"github.com/prometheus/prometheus/promql"
+
+	"github.com/grafana/mimir/pkg/streamingpromql/functions"
+	"github.com/grafana/mimir/pkg/streamingpromql/limiting"
+	"github.com/grafana/mimir/pkg/streamingpromql/types"
+)
+
+type CountAggregationGroup struct {
+	floatValues  []float64
+	floatPresent []bool
```
[the remark this answers is not in the document]
Comment (jhesketh) [reply pkg/streamingpromql/aggregations/count.go:18]: Good idea. I had already done something similar for stddev.

File: pkg/streamingpromql/aggregations/count.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+	var floatPoints []promql.FPoint
+	var err error
+	if floatPointCount > 0 {
+		floatPoints, err = types.FPointSlicePool.Get(floatPointCount, memoryConsumptionTracker)
+		if err != nil {
+			return types.InstantVectorSeriesData{}, false, err
+		}
+
+		for i, fv := range g.values {
+			if fv > 0 {
```
Comment (charleskorn) [inline pkg/streamingpromql/aggregations/count.go:65]: Would be good to add a test that confirms that we don't emit points with 0 values when there are no points at that timestamp

Comment (jhesketh) [reply pkg/streamingpromql/aggregations/count.go:65]: That will be covered by `TestCompareVariousMixedMetrics`, but I'll add something explicit

**Review by charleskorn — APPROVED** [review]

**Pull request merged**